R/decks.R

```diff
@@ -0,0 +1,517 @@
+    out <- AnalysisCatalog(crGET(shojiURL(x, "catalogs", "analyses")))
+    return(out[[1]])
+})
+
+setMethod("analysis<-", c("CrunchSlide", "formula"), function(x, value){
+    analyses[[1]] <- value
+})
+
+setMethod("query<-", "CrunchSlide", function(x, value){
+    analysis <- analyses(x)[[1]]
+    query(analysis) <- value
+    return(invisible(x))
+})
+
+setMethod("cubes", "CrunchSlide", function(x) cubes(analyses(x)))
+setMethod("cube", "CrunchSlide", function(x) cube(analyses(x)[[1]]))
+setMethod("displaySettings", "CrunchSlide", function(x) displaySettings(analyses(x)))
+setMethod("displaySettings<-", "CrunchSlide", function(x, value) {
+    displaySettings(analyses(x)) <- value
+    })
+
+# AnalysisCatalog --------------------------------------------------------------
+
+setMethod("[[", "AnalysisCatalog", function (x, i, ...) {
+  getEntity(x, i, Analysis)
+})
+
+setMethod("[[<-", c("AnalysisCatalog", "numeric", "missing", "formula"), function (x, i, j, value) {
+    if (i > length(x)) {
+        halt("Index out of bounds, you can only assign a formula to an existing analysis.")
+    }
+    analysis <- x[[i]]
+    query(analysis) <- value
+    invisible(refresh(x))
+})
+
+setMethod("[[<-", c("AnalysisCatalog", "numeric", "missing", "Analysis"),
+          function(x, i, j, value){
+              if (length(i) > 1) {
+                  #TODO, recurse through i
+              }
+
+              if (i > length(x) + 1) {
+                  #TODO what to do with adding an analysis that's not the next one.
+              }
+              payload <- value@body[c("query", "display_settings", "query_environment")]
+              payload <- wrapEntity(body = payload)
+              if (i <= length(x)) {
+                  url <- names(x@index)[i]
+                  crPATCH(url, body=toJSON(payload))
+              } else {
+                  crPOST(self(x), body = toJSON(payload))
+              }
+              invisible(refresh(x))
+          })
+
+setMethod("[[<-", c("AnalysisCatalog", "numeric", "missing", "list"), function (x, i, j, value) {
+  all_fmla <- vapply(value, function(x) inherits(x, "formula"), logical(1))
+  if (any(!all_fmla)) {
+      halt("Entry ", which(!all_fmla), " is not a formula")
+  }
+  if (length(i) != length(value)) {
+      noun <- if (length(i) == 1) " analysis." else " analyses."
+      halt("Invalid assignment. You tried to assign ", length(value),
+           " formulas to ", length(i), noun)
+  }
+  mapply(function(analysis, fmla){
+      analysis <- fmla
+  }, analysis = x[[i]], fmla = value)
+})
+
+setMethod("cubes", "AnalysisCatalog", function(x) {
+    lapply(seq_along(x@index), function(i) cube(x[[i]]))
+})
+
+setMethod("displaySettings", "AnalysisCatalog", function(x){
+    analyses <- lapply(seq_along(length(x)), function(i) x[[i]])
+    if (length(x) > 1) {
+        warning("Slide has multiple analyses, returning display settings for the first analysis")
```

Comment:
Should this return a list of displaySettings? I think that's probably the most consistent, but since multi-analysis slides are pretty rare I'm not sure if it's more confusing than beneficial.